python/ray/train/jax/jax_trainer.py

```diff
@@ -0,0 +1,205 @@
+    """A Trainer for data parallel jax training.
+
+    This Trainer runs the function ``train_loop_per_worker`` on multiple Ray
+    Actors. These actors already have the necessary jax process group already
+    configured for distributed jax training.
+
+    The ``train_loop_per_worker`` function is expected to take in either 0 or 1
+    arguments:
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            ...
+
+    .. code-block:: python
+
+        def train_loop_per_worker(config: Dict):
+            ...
+
+    If ``train_loop_per_worker`` accepts an argument, then
+    ``train_loop_config`` will be passed in as the argument. This is useful if you
+    want to tune the values in ``train_loop_config`` as hyperparameters.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    the "train" key), then it will be split into multiple dataset
+    shards that can then be accessed by ``session.get_dataset_shard("train")`` inside
+    ``train_loop_per_worker``. All the other datasets will not be split and
+    ``session.get_dataset_shard(...)`` will return the the entire Dataset.
+
+    Note: Here are some details about how to use the ``JaxTrainer`` properly.
+
+    * cluster scaling config:
+        * For the GPU distributed cases, the inner-device communication is handled
+            internally by Jax; the inter-device communication is set up by the
+            ``JaxTrainer``. Therefore, the ``num_workers`` is set to be the number of
+            nodes; and ``num_gpus_per_worker`` is set to be the number of gpus on each
+            node. For the current experimental version, we only support the homogeneous
+            distributed case, i.e. all the nodes have the same number of gpus. This is
+            also the optimal / balanced case in terms of performance.
+        * For the TPU distributed cases, the multi-device distributed training only
+            supports the TPU pods for a) the TPU pods can host up to several thousands
+            of TPU cores; b) the communication speed is also optimal for the TPU pods.
+            Therefore, the ``num_workers`` is set to be the number of TPU-VMs
+            (say ``num_workers=4`` for TPU-pod v2-32); and
+            ``resources_per_worker={"TPU": 1}`` is used to set the TPU resource.
+            Since the TPU resources are not the default resources on Ray,
+            we need to set the ``resources`` when set up the cluster,
+            e.g. ``ray start --resources='{"TPU":1}'``. For more details, please
+            refer to `Ray Jax TPU Examples </train/examples/jax_tpu>`
+            end to end example.
+        * For the CPU distributed cases, the multi-device distributed training
+            is not supported.
+        * the placement group strategy: Since the JaxTrainer will spread the workers
+            across different nodes, the default value of ``placement_strategy``
+            is changed to ``SPREAD``.
```

Comment:
Thanks @amogkam! The reason for not using `STRICT_SPREAD` is because the head node might be one of the nodes with GPUs. (In other words, the driver node is one of the worker nodes. In this case, only `SPREAD` works.)

This is a good question! In the case of `num_workers=2 and num_gpus_per_worker=2`, the user are recommended to use `num_workers=1 and num_gpus_per_worker=4`. 


------

Ideally, we might want a placement group `STRICT_SPREAD` only for the **worker nodes**. But currently, it seems like `SPREAD` is the closest alternative. How do you think?

Reply:
Why can't the head node be used for training as well?